src/allocator.lisp

```diff
@@ -75,8 +75,13 @@ NOTE: Note that the finalizer may close over the allocated vector."))
               :initial-element (cflonum 0)))
 
 (defmethod allocate-vector ((descr lisp-allocation))
-  (values (make-lisp-cflonum-vector (allocation-length descr))
-          #'dummy-finalizer))
+  (handler-case
+      (values (make-lisp-cflonum-vector (allocation-length descr))
+              #'dummy-finalizer)
+    #+sbcl
+    (sb-kernel::heap-exhausted-error ()
+      (error "Cannot allocate enough memory for ~DQ state."
+             (floor (log (allocation-length descr) 2))))))
```

Comment:
I don't think we should infer the "meaning" of the failed allocation request here. Right now `allocate-vector` gets used both by the pure state and density qvm, and in principle would get used for additional simulation modes (e.g. "path integral" simulation). I would suggest lifting this error message to `make-pure-state` and `make-density-matrix-state` respectively.